tag_tracer.go

```diff
@@ -0,0 +1,265 @@
+
+	// GossipSubConnTagDecayAmount is subtracted from decaying tag values at each decay interval.
+	GossipSubConnTagDecayAmount = 1
+
+	// GossipSubConnTagMessageDeliveryCap is the maximum value for the connection manager tags that
+	// track message deliveries.
+	GossipSubConnTagMessageDeliveryCap = 15
+)
+
+// tagTracer is an internal tracer that applies connection manager tags to peer
+// connections based on their behavior.
+//
+// We tag a peer's connections for the following reasons:
+// - Directly connected peers are tagged with GossipSubConnTagValueDirectPeer (default 1000).
+// - Mesh peers are tagged with a value of GossipSubConnTagValueMeshPeer (default 20).
+//   If a peer is in multiple topic meshes, they'll be tagged for each.
+// - For each message that we receive, we bump a delivery tag for peer that delivered the message
+//   first.
+//   The delivery tags have a maximum value, GossipSubConnTagMessageDeliveryCap, and they decay at
+//   a rate of GossipSubConnTagDecayAmount / GossipSubConnTagDecayInterval.
+type tagTracer struct {
+	sync.RWMutex
+
+	cmgr     connmgr.ConnManager
+	msgID    MsgIdFunction
+	decayer  connmgr.Decayer
+	decaying map[string]connmgr.DecayingTag
+	direct   map[peer.ID]struct{}
+
+	// a map of message ids to the set of peers who delivered the message after the first delivery,
+	// but before the message was finished validating
+	nearFirst map[string]map[peer.ID]struct{}
+}
+
+func newTagTracer(cmgr connmgr.ConnManager) *tagTracer {
+	decayer, ok := connmgr.SupportsDecay(cmgr)
+	if !ok {
+		log.Warnf("connection manager does not support decaying tags, delivery tags will not be applied")
+	}
+	return &tagTracer{
+		cmgr:      cmgr,
+		msgID:     DefaultMsgIdFn,
+		decayer:   decayer,
+		decaying:  make(map[string]connmgr.DecayingTag),
+		nearFirst: make(map[string]map[peer.ID]struct{}),
+	}
+}
+
+func (t *tagTracer) Start(gs *GossipSubRouter) {
+	if t == nil {
+		return
+	}
+
+	t.msgID = gs.p.msgID
+	t.direct = gs.direct
+}
+
+func (t *tagTracer) tagPeerIfDirect(p peer.ID) {
+	if t.direct == nil {
+		return
+	}
+
+	// tag peer if it is a direct peer
+	_, direct := t.direct[p]
+	if direct {
+		t.cmgr.TagPeer(p, "pubsub:direct", GossipSubConnTagValueDirectPeer)
+	}
+}
+
+func (t *tagTracer) tagMeshPeer(p peer.ID, topic string) {
+	tag := topicTag(topic)
+	t.cmgr.TagPeer(p, tag, GossipSubConnTagValueMeshPeer)
+}
+
+func (t *tagTracer) untagMeshPeer(p peer.ID, topic string) {
+	tag := topicTag(topic)
+	t.cmgr.UntagPeer(p, tag)
+}
+
+func topicTag(topic string) string {
+	return fmt.Sprintf("pubsub:%s", topic)
+}
+
+func (t *tagTracer) addDeliveryTag(topic string) {
+	if t.decayer == nil {
+		return
+	}
+
+	t.Lock()
+	defer t.Unlock()
+	tag, err := t.decayingDeliveryTag(topic)
+	if err != nil {
+		log.Warnf("unable to create decaying delivery tag: %s", err)
+		return
+	}
+	t.decaying[topic] = tag
+}
+
+func (t *tagTracer) removeDeliveryTag(topic string) {
+	t.Lock()
+	defer t.Unlock()
+	delete(t.decaying, topic)
```

Comment:
Shouldn't we have a function for erasing a decaying tag explicitly?

Reply:
We don't actually use that though. When would we want to remove a tag? Perhaps when we have a rejection with validationFailed?

Reply:
Actually there is a case we do want to delete the tag when we remove it: imaging Joining, Leaving, and then Joining again a topic; that would lead to an error when we try to recreate the tag.
So yeah, you are right, we have to delete it.

Reply:
https://github.com/libp2p/go-libp2p-core/pull/151 and https://github.com/libp2p/go-libp2p-connmgr/pull/72 are merged.

Will wait til you integrate those features here by importing master. Will make a release once you give me the thumbs up.